src/network/network_crypto.h

```diff
@@ -0,0 +1,282 @@
+	 * @return The size, in bytes, of the MACs.
+	 */
+	virtual size_t MACSize() const = 0;
+
+	/**
+	 * Decrypt the given message in-place, validating against the given MAC.
+	 * @param mac The message authentication code (MAC).
+	 * @param message The location of the message to decrypt.
+	 * @return Whether decryption and authentication/validation of the message succeeded.
+	 */
+	virtual bool Decrypt(std::span<std::uint8_t> mac, std::span<std::uint8_t> message) = 0;
+
+	/**
+	 * Encrypt the given message in-place, and write the associated MAC.
+	 * @param mac The location to write the message authentication code (MAC) to.
+	 * @param message The location of the message to encrypt.
+	 */
+	virtual void Encrypt(std::span<std::uint8_t> mac, std::span<std::uint8_t> message) = 0;
+};
+
+
+/**
+ * Callback interface for requests for passwords in the context of network authentication.
+ */
+class NetworkAuthenticationPasswordRequest {
+public:
+	virtual ~NetworkAuthenticationPasswordRequest() {}
+
+	/**
+	 * Reply to the request with the given password.
+	 */
+	virtual void Reply(const std::string &password) = 0;
+};
+
+/**
+ * Callback interface for client implementations to provide the handling of the password requests.
+ */
+class NetworkAuthenticationPasswordRequestHandler : public NetworkAuthenticationPasswordRequest {
+protected:
+	friend class X25519PAKEClientHandler;
+
+	std::string password; ///< The entered password.
+public:
+
+	virtual void Reply(const std::string &password) override;
+
+	/**
+	 * Callback to trigger sending the response for the password request.
+	 */
+	virtual void SendResponse() = 0;
+
+	/**
+	 * Callback to trigger asking the user for the password.
+	 * @param request The request to the user, to which it can reply with the password.
+	 */
+	virtual void AskUserForPassword(std::shared_ptr<NetworkAuthenticationPasswordRequest> request) = 0;
+};
+
+
+/**
+ * Callback interface for server implementations to provide the current password.
+ */
+class NetworkAuthenticationPasswordProvider {
+public:
+	virtual ~NetworkAuthenticationPasswordProvider() {}
+
+	/**
+	 * Callback to return the password where to validate against.
+	 * @return \c std::string_view of the current password; an empty view means no password check will be performed.
+	 */
+	virtual std::string_view GetPassword() const = 0;
+};
+
+/**
+ * Default implementation of the password provider.
+ */
+class NetworkAuthenticationDefaultPasswordProvider : public NetworkAuthenticationPasswordProvider {
+private:
+	const std::string *password; ///< The password to check against.
+public:
+	/**
+	 * Create the provider with the pointer to the password that is to be used. A pointer, so this can handle
+	 * situations where the password gets changed over time.
+	 * @param password The reference to the configured password.
+	 */
+	NetworkAuthenticationDefaultPasswordProvider(const std::string &password) : password(&password) {}
```

Comment:
Just curious: so why not a `string_view`, as that is a "pointer" too?

Reply:
Because the password can change during the lifetime of the provider, which is at least as long as the user gets to enter their password before getting booted. In practice these objects will just survive the whole life of the application.

Remember that a `std::string_view` is essentially a pointer to a begin and an end, and `std::string` is a length and a pointer to some allocated memory with the content of the string.

Imagine the password being 'a' and I change it to 'bb'. At best the view given to the provider contains 'b' as the number of characters does not change and you might be lucky that `std::string` did not reallocate. At worst you're reading a deallocated string and segfault.

Reply:
tldr; `string_view` is also a length, length of password can change. That is fair :)